### PR TITLE
[Merged by Bors] - feat: Estimator typeclass and implementation for Levenshtein distance

### PR DESCRIPTION
**Pull request by kim-em**
---

- [x] depends on: #6096
- [x] depends on: #6097
- [x] depends on: #6115
- [x] depends on: #6117
- [x] depends on: #6118

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/EditDistance/Estimator.lean [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+instance estimator' :
+    Estimator (Thunk.mk fun _ => (levenshtein C xs ys, ys.length))
+      (LevenshteinEstimator' C xs ys) where
+  bound_le e := match e.pre_rev, e.split, e.bound_eq with
+  | [], split, eq => by
+    simp only [List.reverse_nil, List.nil_append] at split
+    rw [e.distances_eq] at eq
+    simp only [List.getElem_eq_get] at eq
+    rw [split] at eq
```
Comment (PatrickMassot) [inline Mathlib/Data/List/EditDistance/Estimator.lean:84]: What's happening here? Can't you reduce this to a single `simp only` or at least a `simp_rw`?

File: Mathlib/Data/List/EditDistance/Estimator.lean [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+and a particular suffix of `ys`.
+
+This bound is (non-strict) monotone as we take longer suffixes of `ys`.
+
+This is an auxiliary definition for the later `LevenshteinEstimator`:
+this variant constructs a lower bound for the pair consisting of
+the Levenshtein distance from `xs` to `ys`,
+along with the length of `ys`.
+-/
+structure LevenshteinEstimator' : Type _ where
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Data/List/EditDistance/Estimator.lean:43]: Indeed! This PR predates that discovery.

[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Data/List/EditDistance/Estimator.lean:43]: There is no point having any universe polymorphism here at all, in fact, so I'm just removing it.

File: Mathlib/Data/List/EditDistance/Estimator.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Data/List/EditDistance/Estimator.lean:84]: Unfortunately not: changing the first `rw` to a `simp_rw` causes the last `rw` to error with a "motive not type correct". `simp` doesn't make any progress on the when used in place of the last `rw`.

Ugly, hey?

File: Mathlib/Data/List/EditDistance/Estimator.lean [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      exact suffixLevenshtein_minimum_le_levenshtein_append _ _ _
+    · exact List.length_le_of_sublist (List.sublist_append_right _ _)
+  improve_spec e := by
+    dsimp [EstimatorData.improve]
+    match e.pre_rev, e.split, e.bound_eq, e.distances_eq with
+    | [], split, eq, _ =>
+      simp only [List.reverse_nil, List.nil_append] at split
+      rw [e.distances_eq] at eq
+      simp only [List.getElem_eq_get] at eq
+      rw [split] at eq
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Data/List/EditDistance/Estimator.lean:103]: As above.

**Comment by PatrickMassot** [comment]
bors merge

**Pull request closed without merging**